tests/lms/config/resources_test.py

```diff
@@ -136,6 +136,32 @@ def test_h_display_name(
             resources.LTILaunch(pyramid_request).h_display_name == expected_display_name
         )
 
+    def test_h_groupid_raises_if_theres_no_tool_consumer_instance_guid(
+        self, lti_params_for, pyramid_request
+    ):
+        lti_params_for.return_value = {}
+        with pytest.raises(
+            HTTPBadRequest,
+            match='Required parameter "tool_consumer_instance_guid" missing from LTI params',
+        ):
+            resources.LTILaunch(pyramid_request).h_groupid
+
+    def test_h_groupid_raises_if_theres_no_context_id(
+        self, lti_params_for, pyramid_request
+    ):
+        lti_params_for.return_value = {"tool_consumer_instance_guid": "test_guid"}
+        with pytest.raises(
+            HTTPBadRequest,
+            match='Required parameter "context_id" missing from LTI params',
+        ):
+            resources.LTILaunch(pyramid_request).h_groupid
+
+    def test_h_groupid(self, lti_launch):
```

Comment:
I'm having trouble working out what this is testing...I looked at the `lti_launch` fixture but it didn't really clear things up for me. I get that `h_groupid` is a property on `lti_launch` resources, but I'm not grokking how the value being tested for here is derived.

Reply:
Given fixed test values as input (`tool_consumer_instance_guid` and `context_id`) the `groupid` gives a fixed ouptut (since it's a deterministic algorithm) and `"group:d55a3c86dd79d390ec8dc6a8096d0943044ea268@TEST_AUTHORITY"` happens to be the output that our test inputs give.

This test isn't really testing the `d55a3c86dd79d390ec8dc6a8096d0943044ea268` part of the output though, that's not the important part It's just 1. Testing the happy path of the function 2. Testing that the output looks like `group:*@TEST_AUTHORITY`. So the assert could use a regex match. But that'd just be harder to write